src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.Tracing;
+
+namespace System.Net.Sockets
+{
+    [EventSource(Name = "System.Net.Sockets")]
+    internal sealed class SocketsTelemetry : EventSource
+    {
+        public static readonly SocketsTelemetry Log = new SocketsTelemetry();
+
+        [NonEvent]
+        public void ConnectStart(Internals.SocketAddress address)
+        {
+            ConnectStart(address.ToString());
+        }
+
+        [NonEvent]
+        public void ConnectStart(EndPoint address)
+        {
+            ConnectStart(address.ToString());
+        }
+
+        [Event(1, Level = EventLevel.Informational)]
+        public void ConnectStart(string? address)
+        {
+            if (IsEnabled(EventLevel.Informational, EventKeywords.All))
+            {
+                WriteEvent(eventId: 1, address ?? "");
+            }
+        }
+
+        [Event(2, Level = EventLevel.Informational)]
+        public void ConnectStop()
+        {
+            ConnectStopInternal();
+        }
+
+        [Event(3, Level = EventLevel.Error)]
+        public void ConnectFailed(SocketError error, Exception? exception)
+        {
+            if (IsEnabled(EventLevel.Error, EventKeywords.All))
+            {
+                string message = exception?.Message ?? string.Empty;
+                WriteEvent(eventId: 3, (int)error, message);
+                ConnectStopInternal();
+            }
+        }
+
+        [Event(4, Level = EventLevel.Warning)]
+        public void ConnectCanceled()
+        {
+            if (IsEnabled(EventLevel.Warning, EventKeywords.All))
+            {
+                WriteEvent(eventId: 4);
+                ConnectStopInternal();
+            }
+        }
+
+        [NonEvent]
+        private void ConnectStopInternal()
+        {
+            if (IsEnabled(EventLevel.Informational, EventKeywords.All))
+            {
+                WriteEvent(eventId: 2);
+            }
+        }
+    }
+}
```

Comment:
Please add at least basic tests.

Reply:
I could add them, but they were considered brittle and [it was agreed](https://github.com/dotnet/runtime/pull/37619) that telemetry logic is tested manually, thus other completed telemetry PRs do not have them.

>No automated tests were added at this time. The current discussions with others (reverse proxy
team) is that automated tests are brittle in CI. But manual tests will be run.

Reply:
I disagree with that decision.  What's brittle is shipping important code with zero coverage of future changes breaking it.

We have some basic logging tests in Sockets, and I can't remember them every failing.
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Sockets/tests/FunctionalTests/LoggingTest.cs

It should not be hard to write a basic test that validates ConnectStart/Stop events are output.

Reply:
Ok, I will add them.

Reply:
Done.